File: src/03/day3.py
from collections import namedtuple

# symbols: list[int], numbers: dict[int,int]
LineResult = namedtuple('LineResult', ['symbols', 'numbers'])

def parse_line1(line: str) -> list[LineResult]:
    symbols: list[int] = []
    numbers: dict[int,int] = {}
    current_number = 0
    current_index = None
    for i in range(len(line)):
        match line[i]:
            case '.':
                if current_number != 0:
                    numbers[current_index] = current_number
                    current_number = 0
                    current_index = None
            case c if c.isnumeric():
                if current_number == 0:
                    current_index = i
                current_number = current_number * 10 + int(c)
            case _:
                if current_number != 0:
                    numbers[current_index] = current_number
                    current_number = 0
                    current_index = None
                symbols.append(i)
    if current_number != 0:
        numbers[current_index] = current_number
    return LineResult(symbols, numbers)

def parse_line2(line: str) -> list[LineResult]:
    gears: list[int] = []
    numbers: dict[int,int] = {}
    current_number = 0
    current_index = None
    for i in range(len(line)):
        match line[i]:
            case '.':
                if current_number != 0:
                    numbers[current_index] = current_number
                    current_number = 0
                    current_index = None
            case c if c.isnumeric():
                if current_number == 0:
                    current_index = i
                current_number = current_number * 10 + int(c)
            case '*':
                if current_number != 0:
                    numbers[current_index] = current_number
                    current_number = 0
                    current_index = None
                gears.append(i)
            case _:
                if current_number != 0:
                    numbers[current_index] = current_number
                    current_number = 0
                    current_index = None
    if current_number != 0:
        numbers[current_index] = current_number
    return LineResult(gears, numbers)

def part1(filename: str):
    digits = lambda s: len(str(s))
    running_sum = 0
    before = LineResult([], {})
    current = LineResult([], {})
    after = LineResult([], {})
    with open(filename, 'rt') as f:
        for line in f:
            after = parse_line1(line.strip())
            adjacencies = set()
            for index in current.numbers.keys():
                if any(index - 1 <= x <= index + digits(current.numbers[index]) for x in after.symbols):
                    adjacencies.add(index)
                if any(index - 1 <= x <= index + digits(current.numbers[index]) for x in current.symbols):
                    adjacencies.add(index)
                if any(index - 1 <= x <= index + digits(current.numbers[index]) for x in before.symbols):
                    adjacencies.add(index)
            running_sum += sum(current.numbers[i] for i in adjacencies)
            before = current
            current = after
        adjacencies = set()
        for index in current.numbers.keys():
            if any(index - 1 <= x <= index + digits(current.numbers[index]) for x in current.symbols):
                adjacencies.add(index)
            if any(index - 1 <= x <= index + digits(current.numbers[index]) for x in before.symbols):
                adjacencies.add(index)
        running_sum += sum(current.numbers[i] for i in adjacencies)
    print(running_sum)

def part2(filename: str):
    digits = lambda s: len(str(s))
    running_sum = 0
    before = LineResult([], {})
    current = LineResult([], {})
    after = LineResult([], {})
    with open(filename, 'rt') as f:
        for line in f:
            after = parse_line2(line.strip())
            for index in current.symbols:
                adjacencies = []
                adjacencies.extend(value for key, value in before.numbers.items() if key - 1 <= index <= key + digits(value))
                adjacencies.extend(value for key, value in current.numbers.items() if key - 1 <= index <= key + digits(value))
                adjacencies.extend(value for key, value in after.numbers.items() if key - 1 <= index <= key + digits(value))
                if len(adjacencies) == 2:
                    running_sum += adjacencies[0] * adjacencies[1]
            before = current
            current = after
        for index in current.symbols:
            adjacencies = []
            adjacencies.extend(value for key, value in before.numbers.items() if key - 1 <= index <= key + digits(value))
            adjacencies.extend(value for key, value in current.numbers.items() if key - 1 <= index <= key + digits(value))
            if len(adjacencies) == 2:
                print(adjacencies)
                running_sum += adjacencies[0] * adjacencies[1]
    print(running_sum)

if __name__ == '__main__':
    part1('day3_input.txt')
    part2('day3_input.txt')

File: src/03/day3.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs;

pub struct LineResult {
    pub symbols: Vec<i32>,
    pub numbers: HashMap<i32, i32>
}

impl LineResult {
    pub fn new() -> LineResult {
        return LineResult {
            symbols: Vec::new(),
            numbers: HashMap::new()
        }
    }
}

fn parse_line(line: &str, mode: bool) -> LineResult {
    let mut symbols: Vec<i32> = Vec::new();
    let mut numbers: HashMap<i32, i32> = HashMap::new();

    let mut current_number: i32 = 0;
    let mut current_index: Option<i32> = None;

    let mut index: i32 = 0;

    for character in line.chars() {
        match character {
            '.' => {
                match current_index {
                    Some(idx) => {
                        numbers.insert(idx, current_number);
                        current_number = 0;
                        current_index = None;
                    },
                    None => {}
                }
            },
            c if c.is_numeric() => {
                match current_index {
                    Some(_) => {},
                    None => {current_index = Some(index)}
                }
                current_number = current_number * 10 +
                                 c.to_digit(10).unwrap() as i32;
            },
            '*' if mode => {
                match current_index {
                    Some(index) => {
                        numbers.insert(index, current_number);
                        current_number = 0;
                        current_index = None;
                    },
                    None => {}
                }
                symbols.push(index);
            },
            _ => {
                match current_index {
                    Some(index) => {
                        numbers.insert(index, current_number);
                        current_number = 0;
                        current_index = None;
                    },
                    None => {}
                }
                if !mode {
                    symbols.push(index);
                }
            }
        }
        index += 1;
    }
    if current_number != 0 {
        numbers.insert(current_index.unwrap(), current_number);
    }
    return LineResult {
        symbols: symbols,
        numbers: numbers
    };
}

fn part1(filename: &str) -> i32 {
    let digits = |s| {
        let mut x = s;
        let mut i: i32 = 0;
        while x > 0 {
            x /= 10;
            i += 1;
        }
        return i;
    };

    let mut running_sum: i32 = 0;

    let mut before = LineResult::new();
    let mut current = LineResult::new();
    let mut after;

    let mut adjacencies: HashSet<i32> = HashSet::new();

    let lines = fs::read_to_string(filename).expect("File not found");

    for line in lines.split_whitespace().map(|x| x.trim()) {
 
        after = parse_line(&line, false);
        adjacencies.clear();

        for (index, number) in &current.numbers {

            let adjacent = |&x| {*index - 1 <= x && x <= *index + digits(*number)};

            if after.symbols.iter().any(adjacent) {
                adjacencies.insert(*index);
            }
            if current.symbols.iter().any(adjacent) {
                adjacencies.insert(*index);
            }
            if before.symbols.iter().any(adjacent) {
                adjacencies.insert(*index);
            }
        }
        for index in &adjacencies {
            running_sum += current.numbers.get(&index).unwrap();
        }
        before = current;
        current = after;
    }

    adjacencies.clear();
    for (index, number) in &current.numbers {

        let adjacent = |&x| {*index - 1 <= x && x <= *index + digits(*number)};

        if current.symbols.iter().any(adjacent) {
            adjacencies.insert(*index);
        }
        if before.symbols.iter().any(adjacent) {
            adjacencies.insert(*index);
        }
    }
    for index in &adjacencies {
        running_sum += current.numbers.get(&index).unwrap();
    }

    return running_sum;
}

fn part2(filename: &str) -> i32 {
    let digits = |s| {
        let mut x = s;
        let mut i: i32 = 0;
        while x > 0 {
            x /= 10;
            i += 1;
        }
        return i;
    };

    let mut running_sum: i32 = 0;

    let mut before = LineResult::new();
    let mut current = LineResult::new();
    let mut after;

    let mut adjacencies: Vec<i32> = Vec::new();

    let lines = fs::read_to_string(filename).expect("File not found");

    for line in lines.split_whitespace().map(|x| x.trim()) {

        after = parse_line(&line, true);

        for index in &current.symbols {

            let adjacent_value = |(&key, &value)| {
                if key - 1 <= *index && *index <= key + digits(value) {Some(value)} else {None}
            };

            adjacencies.clear();

            adjacencies.extend(before.numbers.iter().filter_map(adjacent_value));
            adjacencies.extend(current.numbers.iter().filter_map(adjacent_value));
            adjacencies.extend(after.numbers.iter().filter_map(adjacent_value));

            if adjacencies.len() == 2 {
                running_sum += adjacencies[0] * adjacencies[1];
            }
        }
        before = current;
        current = after;
    }
    for index in &current.symbols {

        let adjacent_value = |(&key, &value)| {
            if key - 1 <= *index && *index <= key + digits(value) {Some(value)} else {None}
        };

        adjacencies.clear();

        adjacencies.extend(before.numbers.iter().filter_map(adjacent_value));
        adjacencies.extend(current.numbers.iter().filter_map(adjacent_value));

        if adjacencies.len() == 2 {
            running_sum += adjacencies[0] * adjacencies[1];
        }
    }

    return running_sum;
}

fn main() {
    println!("{}", part1("day3_input.txt"));
    println!("{}", part2("day3_input.txt"));
}